File: 0040-combination-sum-ii/0040-combination-sum-ii.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function (candidates, target) {
  const answer = [];

  candidates.sort((a, b) => a - b);

  const findSum = (index, currTarget, arr) => {
    if (currTarget === 0) {
      answer.push(arr.slice());
      return;
    }
    
    if (index >= candidates.length || currTarget < 0) return;

    arr.push(candidates[index]);

    findSum(index + 1, currTarget - candidates[index], arr);
    arr.pop();

    let nextIndex = index + 1;
    while (nextIndex < candidates.length && candidates[nextIndex] === candidates[index]) nextIndex++;
    findSum(nextIndex, currTarget, arr);
  }

  findSum(0, target, []);

  return answer;
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
  def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
    answer = []
      
    def findSum(index, arr, currTarget):
      if currTarget == 0:
        answer.append(list(arr))
        return

      if index >= len(candidates) or currTarget < 0:
            return

      arr.append(candidates[index])

      findSum(index + 1, arr, currTarget - candidates[index])
      arr.pop()
      nextIndex = index + 1
      while(nextIndex < len(candidates) and candidates[nextIndex] == candidates[index]):
        nextIndex += 1
      findSum(nextIndex, arr, currTarget)

    candidates.sort()
    findSum(0, [], target)
    return answer